### PR TITLE
[lesson5] added api tool for tables reservations

### PR DESCRIPTION
**Pull request by wizzardio**

### CODE REVIEW

File: lesson5_api_tool/controllers/Reservations.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  /**
+   * Find reservation by table ID
+   * @param {String} table - table`s ID
+   * @return {Object} - reservation or null
+   */
+  static async findReservationByTableId(table) {
+    try {
+      const res = await db.select('reservations.*').from(RESERVATION_TABLE)
```
Comment (webwin0) [inline lesson5_api_tool/controllers/Reservations.js]: You can use https://knexjs.org/#Builder-first

File: lesson5_api_tool/controllers/Reservations.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+class Reservations {
+  /**
+   * Created new reservation.
+   * @param {Object} data - data for creation.
+   * This param should contain:
+   * 1) guests - amount of guests;
+   * 2) start - start of reservation;
+   * 3) end - end of reservation;
+   * 4) table - table`s ID;
```
Comment (webwin0) [inline lesson5_api_tool/controllers/Reservations.js]: `table` is confusing name, better `tableId`

File: lesson5_api_tool/knexfile.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+
+const DB = 'postgres://postgres:1941233@localhost:5432/node_lesson5'; // not the best way to store DB creds
```
Comment (webwin0) [inline lesson5_api_tool/knexfile.js]: You can use https://www.npmjs.com/package/dotenv for getting configs from environment variables

File: lesson5_api_tool/routes/api/reservations.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      guests, start: time, end: to, table: freeTable.id,
+    });
+    logger.info(JSON.stringify(result));
+    const reservation = await ReservationsCTRL.findReservationByTableId(freeTable.id);
+    if (!reservation) {
+      return res.status(404).json({ msg: `Was not able to find reservation for table [${freeTable.id}]` });
+    }
+    return res.json({ location: `/api/reservations/${reservation.id}` });
+  } catch (err) {
+    return res.status(400).json({ err: `Unexpected error while tried to save reservation for this data [${JSON.stringify(req.body)}]...` });
```
Comment (webwin0) [inline lesson5_api_tool/routes/api/reservations.js]: http code should be 500

File: lesson5_api_tool/routes/api/reservations.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }
+    const result = await ReservationsCTRL.createReservation({
+      guests, start: time, end: to, table: freeTable.id,
+    });
+    logger.info(JSON.stringify(result));
+    const reservation = await ReservationsCTRL.findReservationByTableId(freeTable.id);
+    if (!reservation) {
+      return res.status(404).json({ msg: `Was not able to find reservation for table [${freeTable.id}]` });
+    }
+    return res.json({ location: `/api/reservations/${reservation.id}` });
```
Comment (webwin0) [inline lesson5_api_tool/routes/api/reservations.js]: .status(201)

File: lesson5_api_tool/routes/api/reservations.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  let updateRes = false;
+  try {
+    const reservation = await ReservationsCTRL.getReservation(req.params.id);
+    if (!reservation) {
+      return res.status(404).json({ msg: `Was not able to find reservation with ID [${req.params.id}]` });
+    }
+    const freeTable = await TablesCTRL.getFreeTable(guests, time, to);
+    if (!freeTable) {
+      return res.status(400).json({ msg: `There are not free table for [${time}]` });
```
Comment (webwin0) [inline lesson5_api_tool/routes/api/reservations.js]: 404

File: lesson5_api_tool/routes/api/reservations.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    reservation.start = time;
+    reservation.end = to;
+    reservation.guests = guests;
+    updateRes = await ReservationsCTRL.updateReservation(reservation);
+    if (!updateRes) {
+      return res.status(409).json({ err: `Unable to modify reservation ${reservation.id}...` });
+    }
+    return res.json({ location: `/api/reservations/${reservation.id}` });
+  } catch (err) {
+    return res.status(400).json(err.message);
```
Comment (webwin0) [inline lesson5_api_tool/routes/api/reservations.js]: 500

File: lesson5_api_tool/routes/api/reservations.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  }
+});
+
+router.delete('/:id', async (req, res) => {
+  try {
+    const status = await ReservationsCTRL.deleteReservation(req.params.id);
+    return res.status(status ? 204 : 404).json();
```
Comment (webwin0) [inline lesson5_api_tool/routes/api/reservations.js]: return res.status(status ? 204 : 404).end();

File: lesson5_api_tool/routes/api/reservations.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  } catch (err) {
+    return res.status(400).json({ err: `Unexpected error while tried to delete reservation with ID [${req.params.id}]` });
+  }
+});
+
+router.get('/:id', async (req, res) => {
+  try {
+    const result = await ReservationsCTRL.getReservation(req.params.id);
```
Comment (webwin0) [inline lesson5_api_tool/routes/api/reservations.js]: Validation is required for `id` param. You can add it in middleware

File: lesson5_api_tool/utils/validate.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+const moment = require('moment'); // not the best idea to add all lib for one verification
+
+const validate = (guests, time, duration) => {
+  const errors = {};
+  const dateFormat = 'YYYY-MM-DDTHH:mm:ssZ';
+
+  if (typeof guests !== 'number') {
+    errors.guests = `Guests value should be number, you entered ${typeof guests}`;
+  } else if (guests <= 1 || guests >= 10) {
```
Comment (webwin0) [inline lesson5_api_tool/utils/validate.js]: guests > 10

Comment (wizzardio) [reply lesson5_api_tool/utils/validate.js]: done

File: lesson5_api_tool/routes/api/reservations.js (hunk 8 above)
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: done

File: lesson5_api_tool/routes/api/reservations.js (hunk 6 above)
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: done

File: lesson5_api_tool/routes/api/reservations.js (hunk 5 above)
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: done

File: lesson5_api_tool/routes/api/reservations.js (hunk 4 above)
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: done

File: lesson5_api_tool/routes/api/reservations.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  const { errors, isValid } = validate(guests, time, duration);
+  if (!isValid) {
+    return res.status(400).json(errors);
+  }
+  const to = new Date(dateFrom.setTime(dateFrom.getTime() + (duration * 60 * 60 * 1000)));
+
+  try {
+    const freeTable = await TablesCTRL.getFreeTable(guests, time, to);
+    if (!freeTable) {
+      return res.status(400).json({ msg: `There are not free table for [${time}]` });
```
[the remark this answers is not in the document]
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: agree.

File: lesson5_api_tool/controllers/Reservations.js (hunk 2 above)
Comment (wizzardio) [reply lesson5_api_tool/controllers/Reservations.js]: done

File: lesson5_api_tool/controllers/Reservations.js (hunk 1 above)
Comment (wizzardio) [reply lesson5_api_tool/controllers/Reservations.js]: done

File: lesson5_api_tool/knexfile.js (hunk 3 above)
Comment (wizzardio) [reply lesson5_api_tool/knexfile.js]: agree, thanks for the hint

File: lesson5_api_tool/routes/api/reservations.js (hunk 7 above)
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: done

File: lesson5_api_tool/routes/api/reservations.js (hunk 9 above)
Comment (wizzardio) [reply lesson5_api_tool/routes/api/reservations.js]: done

**Comment by wizzardio** [comment]
@webwin0 thanks for code review and sorry for silly mistakes. 
I'm overloaded with JS and Node right now)
Promise to refactor all this staff for the next HM.

**Pull request closed without merging**